### PR TITLE
Allow the ability to set shareNativeConnection to false for LettuceConnectionFactory

### PR DESCRIPTION
**Pull request by ayudovin**
Allow the ability to set shareNativeConnection for LettuceConnectionFactory
this [enhancement](https://github.com/spring-projects/spring-boot/issues/14196)

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 1]
```diff
@@ -76,7 +78,15 @@ public LettuceConnectionFactory redisConnectionFactory(
 			ClientResources clientResources) throws UnknownHostException {
 		LettuceClientConfiguration clientConfig = getLettuceClientConfiguration(
 				clientResources, this.properties.getLettuce().getPool());
-		return createLettuceConnectionFactory(clientConfig);
+		LettuceConnectionFactory lettuceConnectionFactory = createLettuceConnectionFactory(
+				clientConfig);
+
+		if (Objects.nonNull(this.properties.getLettuce().getShareNativeConnection())) {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java]: This looks odd that this property alone has to be handled separately.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java [hunk 2]
```diff
@@ -325,6 +326,14 @@ public void setPool(Pool pool) {
 		 */
 		private Pool pool;
 
+		/**
+		 * If shareNativeConnection is true, the pool will be used to select a connection
+		 * for blocking and tx operations only, which should not share a connection. If
+		 * native connection sharing is false, the selected connection will be used for
+		 * all operations
+		 */
+		private Boolean shareNativeConnection;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java]: We should make up our mind about a default if that makes sense. If it does, it should be reflected here (and tested).

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java [hunk 3]
```diff
@@ -325,6 +326,14 @@ public void setPool(Pool pool) {
 		 */
 		private Pool pool;
 
+		/**
+		 * If shareNativeConnection is true, the pool will be used to select a connection
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java]: That's a bit unusual for the description of a key (see other descriptions around for inspiration). The first sentence must be a concise summary of the purpose of the key as it's the one showing up in auto-completion popup in an IDE.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/redis/RedisAutoConfigurationTests.java [hunk 4]
```diff
@@ -150,12 +151,15 @@ public void testPasswordInUrlStartsWithColon() {
...
 	@Test
 	public void testRedisConfigurationWithPool() {
-		this.contextRunner.withPropertyValues("spring.redis.host:foo",
-				"spring.redis.lettuce.pool.min-idle:1",
-				"spring.redis.lettuce.pool.max-idle:4",
-				"spring.redis.lettuce.pool.max-active:16",
-				"spring.redis.lettuce.pool.max-wait:2000",
-				"spring.redis.lettuce.shutdown-timeout:1000").run((context) -> {
+		this.contextRunner
+				.withPropertyValues("spring.redis.host:foo",
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/redis/RedisAutoConfigurationTests.java]: I am not sure that reformat is necessary for adding a single key

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java (hunk 2 above)
Comment (ayudovin) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisProperties.java]: the default value of  `shareNativeConnection` is true.  This default value is located in `LettuceConnectionFactory` but we also could the same default value in `RedisProperties`

**Comment by mp911de** [comment]
`LettuceClientConfiguration` encapsulates driver-specifics. Connection sharing and other properties like connection validation are specific to `LettuceConnectionFactory`.

Changing connection sharing settings is a somewhat rare use-case and I'd rather propose to use a `BeanPostProcessor` on the application side to tweak settings where necessary. I left a comment in #14196 why I think using multiple connections won't really help.

**Comment by snicoll** [comment]
Thanks for the feedback @mp911de 

@ayudovin thanks for the PR anyway.

**Pull request closed without merging**